kedro_mlflow/io/mlflow_metrics_dataset.py

```diff
@@ -0,0 +1,178 @@
+from copy import deepcopy
+from functools import partial, reduce
+from itertools import chain
+from typing import Any, Dict, Generator, List, Optional, Tuple, Union
+
+import mlflow
+from kedro.io import AbstractDataSet, DataSetError
+from mlflow.tracking import MlflowClient
+
+MetricItem = Union[float, List[float], Dict[str, float], List[Dict[str, float]]]
+MetricTuple = Tuple[str, float, int]
+MetricsDict = Dict[str, MetricItem]
+
+
+class MlflowMetricsDataSet(AbstractDataSet):
+    """This class represent MLflow metrics dataset."""
+
+    def __init__(
+        self,
+        run_id: str = None,
+        prefix: Optional[str] = None,
+    ):
+        """Initialise MlflowMetricsDataSet.
+
+        Args:
+            prefix (Optional[str]): Prefix for metrics logged in MLflow.
+            run_id (str): ID of MLflow run.
+        """
+        self._prefix = prefix
+        self._run_id = run_id
+
+    def _load(self) -> MetricsDict:
+        """Load MlflowMetricDataSet.
+
+        Returns:
+            Dict[str, Union[int, float]]: Dictionary with MLflow metrics dataset.
+        """
+        client = MlflowClient()
+        run_id = self._get_run_id()
+        all_metrics = client._tracking_client.store.get_all_metrics(run_uuid=run_id)
+        dataset_metrics = filter(self._is_dataset_metric, all_metrics)
+        dataset = reduce(
+            lambda xs, x: self._update_metric(
+                # get_all_metrics returns last saved values per metric key.
+                # All values are required here.
+                client.get_metric_history(run_id, x.key),
+                xs,
+            ),
+            dataset_metrics,
+            {},
+        )
+        return dataset
```

Comment:
Loading will always give a value with formatting for `"metric4": [{"value": 1.1, "step": 1}, {"value": 1.2, "step": 2}]`, and a sa consequence, `x!=load(save(x))` (when a user has saved data with format 1,2 or3 above). This may be very confusing. Should we force the user to always provide this verbose format? My fear is that if you want to reuse data from one node to another in a kedro pipeline, it is totally expected that the node1 inputs is totally identical to node2 output.

Does it makes sense (I know that I previously suggested to be able to log with any of these 4 format...)? At least we should log a warning when saving

Reply:
Yes, it makes a lot of sense. It simplifies interaction with the API and code, so definitely that approach is better.

Reply:
Another method would be to keep the different ways of saving, and add a `save_args`parameter to specify if we want to get the `last` value of the metric, the `list` (i.e. the list of all values) or the history (dict with step and values). it will cover more use cases, but it can be a bit confusing because it adds parameterization.